### DISCUSSION TITLE
Add sound filters to PC Speaker

### DISCUSSION THREAD

**Discussion by Fenix770** [86Box/86Box#3053] [Ideas]
**Is your feature request related to a problem? Please describe.**
I would like the implementation of PC speaker sound filters like those IBMulator have. The reason is the old games that use the
PC speaker sounds more realistic with these filters.

**Describe the solution you'd like**
I don't know if it's possible to port these filters from IBMulator to 86box.

**Additional context**
This is the filter options the IBMulator have inside ibmulator.ini file:

_[pcspeaker]
; enabled: Enable PC-Speaker emulation.
;    rate: Sample rate.
;          Possible values: 48000, 44100, 32000, 22050, 11025.
; filters: DSP filters. Use them to emulate the response of the typical PC speaker.
;          Possible values: a list of filter definitions. See the README for more info.
;          Example: LowPass,order=5,cutoff=5000|HighPass,order=5,cutoff=500
;  volume: Audio volume.
enabled=yes
rate=48000
filters=LowPass,order=5,cutoff=5000|HighPass,order=5,cutoff=500
volume=0.5_

**Comment by OBattler** [comment]
We do do sound filtering for the PC speaker already, just the parameters need tweaking.

**Comment by usergithub64** [comment]
So you can add a volume control for the PC Speaker?

**Comment by OBattler** [comment]
I already explained that I tried to do that and it didn't work.

**Comment by Ompronce** [comment]
Are the characteristics of a piezoelectric speaker emulatable, compared to a cone speaker (which I assume is what's currently emulated)?

**Comment by OBattler** [comment]
From what I know, a piezoelectric speaker is instant response, so all the fancy stuff, such as the ability to play waveform audio on the PC speaker, would not be possible on one.

**Comment by Ohrwerk** [comment]
There were games that had an " understandable " sound quality through the PC speaker.
The sounds that the computer could produce were changed so quickly that "intermediate sounds" were created that were not supported by the hardware.
But this only worked with a real PC speaker, the piezo beepers couldn't do that.

If you started Silent Service 2 with the IBM (PC Speaker) option. When starting the game Digital Speech Output answered Yes, there was no music, but the speech output was through the PC Speaker and you could understand the words.

But the option to disable the PC speaker would be nice.
Could not find anything on such a possibility.
[pcspeaker]
enabled=false

**Comment by dbalsom** [comment]
There's a bit of false folklore surrounding the PC speaker and its special ability to play PCM audio due to 'intermediate states' or whathaveyou. It's not true for several reasons; speakers were different sizes, and different materials, and certainly different response curves. If programmers had to account for such delicate effects you'd have to pick your speaker type when installing the game.

When I started implementing PC speaker audio in MartyPC I played around with various models of speaker response, and really nothing made the resulting audio sound any better than just resampling the PIT output and playing it directly. The IBM PC and XT do however pass the PIT output through a simple lowpass filter before sending it to the speaker, which it seems like what IBMulator is simulating.  There's sort of a cheap way of doing this by just disabling speaker output if the frequency of PIT channel 2 is too high. 

The notion that you can't get PCM audio out of a piezo is incorrect. It's important to note there is a difference between a piezo speaker and a piezo beeper -  a beeper is pretty much locked to a specific tone, and of course that's useless. But there were systems back in the day with piezo speakers. I uploaded an example of Magic Mushroom running on the Book8088 which has a little piezo speaker near the USB port - it's not quite as loud as on a real membrane speaker, but it's certainly intelligible:

https://www.youtube.com/watch?v=IWBKb2ZOVwc
